src/util/psbt/mod.rs

```diff
@@ -210,6 +213,234 @@ impl PartiallySignedTransaction {
 
         Ok(())
     }
+
+    /// Signs the input at `input_index` with private key `sk`.
+    ///
+    /// # Errors
+    /// If an attempt is made to sign a taproot input.
+    pub fn sign<C: Signing>(&mut self, sk: &PrivateKey, input_index: usize, secp: &Secp256k1<C>) -> Result<(), SighashError> {
+        self.check_index_is_within_bounds(input_index)?;
+
+        match self.signing_algorithm(input_index)? {
+            Algo::Ecdsa => self.sign_ecdsa(sk, input_index, secp),
+            Algo::Schnorr => Err(SighashError::TaprootNotYetSupported),
+        }
+    }
+
+    fn sign_ecdsa<C: Signing>(&mut self, sk: &PrivateKey, input_index: usize, secp: &Secp256k1<C>) -> Result<(), SighashError> {
+        let mut cache = SighashCache::new(&self.unsigned_tx);
```

Comment:
I don't like having the cache as an additional argument. But we are losing the benefits of cache if we don't use reu-use it while signing other inputs.

Reply:
Thanks, I added an optional cache to the `sign` function so that signing multiple inputs is possible with the same cache.